### PR TITLE
support lottery ticket hypothesis

### PR DESCRIPTION
**Pull request by QuanluZhang**
This is an implementation of the paper "The Lottery Ticket Hypothesis: Finding Sparse, Trainable Neural Networks" on NNI.

finished: pytorch version, doc, docstring
todo: tensorflow 2.0 version

### CODE REVIEW

**Comment by QuanluZhang** [comment]
@tanglang96 please help review

File: docs/en_US/Compressor/Overview.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ We have provided two naive compression algorithms and three popular ones for use
 |---|---|
 | [Level Pruner](./Pruner.md#level-pruner) | Pruning the specified ratio on each weight based on absolute values of weights |
 | [AGP Pruner](./Pruner.md#agp-pruner) | Automated gradual pruning (To prune, or not to prune: exploring the efficacy of pruning for model compression) [Reference Paper](https://arxiv.org/abs/1710.01878)|
+| [Lottery Ticket Pruner](./Pruner.md#agp-pruner) | Automated gradual pruning (To prune, or not to prune: exploring the efficacy of pruning for model compression) [Reference Paper](https://arxiv.org/abs/1710.01878)|
```
Comment (tanglang96) [inline docs/en_US/Compressor/Overview.md]: paper name and link not correct

File: src/sdk/pynni/nni/compression/torch/lottery_ticket.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    This is a Pytorch implementation of the paper "The Lottery Ticket Hypothesis: Finding Sparse, Trainable Neural Networks",
+    following NNI model compression interface.
+
+    1. Randomly initialize a neural network f(x;theta_0) (where theta_0 follows D_{theta}).
+    2. Train the network for j iterations, arriving at parameters theta_j.
+    3. Prune p% of the parameters in theta_j, creating a mask m.
+    4. Reset the remaining parameters to their values in theta_0, creating the winning ticket f(x;m*theta_0).
+    5. Repeat step 2, 3, and 4.
+    """
+    def __init__(self, model, config_list, optimizer, reset_weights=True):
```
Comment (tanglang96) [inline src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: maybe lr_scheduler also needs to be reset and it sometimes relies on the `epoch`, maybe init the `LotteryTicketPruner` with a `checkpoint` dict could be better, which contains the `state_dict` of the model, optimizer and scheduler, then prune the model in a loop, like
```
for _ in range(prune_iterations):
    for epoch in range(epoch_num):
        pruner.update_epoch(epoch)
        print('# Epoch {} #'.format(epoch))
        train(model, device, train_loader, optimizer)
```

File: docs/en_US/Compressor/Overview.md (hunk 1 above)
Comment (QuanluZhang) [reply docs/en_US/Compressor/Overview.md]: good catch, thanks.

File: src/sdk/pynni/nni/compression/torch/lottery_ticket.py (hunk 2 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: how to checkpoint lr_scheduler?

File: examples/model_compress/main_torch_pruner.py [hunk 3]
```diff
@@ -96,3 +96,4 @@ def main():
 
 if __name__ == '__main__':
     main()
+
```
Comment (liuzhe-lz) [inline examples/model_compress/main_torch_pruner.py]: What's this?
Maybe some incorrect `CRLF` like stuff?

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 4]
```diff
@@ -204,6 +202,9 @@ def export_model(self, model_path, mask_path=None, onnx_path=None, input_shape=N
         input_shape : list or tuple
             input shape to onnx model
         """
+        if self.detect_modules_to_compress() and not self.mask_dict:
+            _logger.warning('You may not use self.mask_dict in base Pruner class to record masks')
+            print('Warning: You may not use self.mask_dict in base Pruner class to record masks')
```
Comment (liuzhe-lz) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: Please remove the print.
Maybe we should add logger configuration code for "standalone mode" now.

File: src/sdk/pynni/nni/compression/torch/lottery_ticket.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+    def calc_mask(self, layer, config, **kwargs):
+        """
+        Generate mask for the given ``weight``.
+
+        Parameters
+        ----------
+        layer : LayerInfo
+            The layer to be pruned
+        config: dict
```
Comment (liuzhe-lz) [inline src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: Space.

File: src/sdk/pynni/nni/compression/torch/lottery_ticket.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        return range(self.prune_iterations + 1)
+
+    def prune_iteration_start(self):
+        """
+        Control the pruning procedure on updated epoch number.
+        Should be called at the beginning of the epoch.
+
+        Parameters
+        ----------
+        epoch: num
```
Comment (liuzhe-lz) [inline src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: Space.

File: src/sdk/pynni/nni/compression/torch/lottery_ticket.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        if self.curr_prune_iteration == 0:
+            mask = torch.ones(weight.shape).type_as(weight)
+        else:
+            curr_sparsity = self._calc_sparsity(sparsity)
+            assert self.mask_dict.get(op_name) is not None
+            curr_mask = self.mask_dict.get(op_name)
+            w_abs = weight.abs() * curr_mask
+            k = int(w_abs.numel() * curr_sparsity)
+            threshold = torch.topk(w_abs.view(-1), k, largest=False).values.max()
+            mask = torch.gt(w_abs, threshold).type_as(weight)
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: w_abs.view(-1) mixes all dimensions together.  
For linear layer weight, it looks fine.    For the weights with multiple dimensions such as Conv2d, I am not sure if it work as expected?

File: src/sdk/pynni/nni/compression/torch/lottery_ticket.py (hunk 2 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: good suggestion, refactored

**Comment by liuzhe-lz** [comment]
I have commented on all style issues I found.
But I can't understand the algorithm :(

**Comment by QuanluZhang** [comment]
> I have commented on all style issues I found.
> But I can't understand the algorithm :(

how about reading the docs?

File: src/sdk/pynni/nni/compression/torch/lottery_ticket.py (hunk 7 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: the expected behavior is merging all the dimensions, it does not treat different dimensions differently.

File: src/sdk/pynni/nni/compression/torch/lottery_ticket.py (hunk 5 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: fixed, thanks

File: src/sdk/pynni/nni/compression/torch/compressor.py (hunk 4 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/torch/compressor.py]: any proposal?

File: src/sdk/pynni/nni/compression/torch/lottery_ticket.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        return mask
+
+    def calc_mask(self, layer, config, **kwargs):
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: I remember this **kwargs is removed?

Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: yes, you are right!

Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: removed, thanks

File: src/sdk/pynni/nni/compression/torch/lottery_ticket.py (hunk 7 above)
Comment (chicm-ms) [reply src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: Got it, thanks.

File: src/sdk/pynni/nni/compression/torch/lottery_ticket.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        Generate mask for the given ``weight``.
+
+        Parameters
+        ----------
+        layer : LayerInfo
+            The layer to be pruned
+        config : dict
+            Pruning configurations for this weight
+        **kwargs
+            Not used
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: This kwargs docstring can be remove, too.

File: src/sdk/pynni/nni/compression/torch/lottery_ticket.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+    def prune_iteration_start(self):
+        """
+        Control the pruning procedure on updated epoch number.
+        Should be called at the beginning of the epoch.
+
+        Parameters
+        ----------
+        epoch : num
+            The current epoch number provided by user call
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: there is no 'epoch' parameter.

Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: remove, thanks!

File: src/sdk/pynni/nni/compression/torch/lottery_ticket.py (hunk 9 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/torch/lottery_ticket.py]: removed, thanks

File: src/sdk/pynni/nni/compression/torch/compressor.py (hunk 4 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/torch/compressor.py]: removed, and add a new issue to track this log refactor

File: examples/model_compress/lottery_torch_mnist.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
+from nni.compression.torch import AGP_Pruner, LevelPruner, LotteryTicketPruner
```
Comment (tanglang96) [inline examples/model_compress/lottery_torch_mnist.py]: maybe keep one of the ```lottery_torch_mnist.py``` or ```lottery_torch_mnist_fc.py``` is enough

File: examples/model_compress/lottery_torch_mnist_fc.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        accuracy = 100. * correct / len(test_loader.dataset)
+    return accuracy
+
+
+if __name__ == '__main__':
+    transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))])
+    traindataset = datasets.MNIST('./data', train=True, download=True, transform=transform)
+    testdataset = datasets.MNIST('./data', train=False, transform=transform)
+    train_loader = torch.utils.data.DataLoader(traindataset, batch_size=60, shuffle=True, num_workers=0, drop_last=False)
+    test_loader = torch.utils.data.DataLoader(testdataset, batch_size=60, shuffle=False, num_workers=0, drop_last=True)
```
Comment (tanglang96) [inline examples/model_compress/lottery_torch_mnist_fc.py]: why ```drop_last=True``` in the test_loader?

File: examples/model_compress/lottery_torch_mnist.py (hunk 11 above)
Comment (QuanluZhang) [reply examples/model_compress/lottery_torch_mnist.py]: let's keep it for now.

File: examples/model_compress/lottery_torch_mnist_fc.py (hunk 12 above)
Comment (QuanluZhang) [reply examples/model_compress/lottery_torch_mnist_fc.py]: it is okay to drop the last.
